Updating dev to current project state

Kylep342 opened the pull request.

Kylep342 committed:
Merge pull request #1 from Kylep342/dev

Dev

Kylep342 committed:
Merge pull request #2 from Kylep342/dev

Dev

Kylep342 committed:
Merge pull request #3 from Kylep342/dev

Set up example 'config.yml' file

Kylep342 committed:
Merge pull request #4 from Kylep342/dev

Dev

Kylep342 committed:
Merge pull request #5 from Kylep342/dev

Dev

Kylep342 committed:
Merge pull request #6 from Kylep342/dev

Removing ubuntu setup from Dockerfile

Kylep342 committed:
Moving from build-time .yml configuration to runtime env (.list) configuration

Kylep342 committed:
Restructuring as app/package architecture
Added a .dockerignore to ignore __pycache__
Renamed thorcast() function to lookup() to remove name confusion

Kylep342 committed:
Removing apikey as function argument for geocode; is now explicit within the code

Kylep342 committed:
Simple Flask API set up

Kylep342 committed:
Updated README to have an example for Flask API setup

Kylep342 committed:
Reformatting README

Kylep342 committed:
work on backing API with Postgresql database to store coordinates for cities that have been previously requested

Kylep342 committed:
Modifications made to coords.py to guard against sql injection
Added more files to .dockerignore
Added database directory and scripts to set up tables needed for db
Added utils folder to hold module for future helper functions

Kylep342 committed:
Merge pull request #7 from Kylep342/memoize_coordinates

Memoize coordinates

Kylep342 committed:
Reorganizing server code across files
Adding some docstrings

Kylep342 committed:
Modifying docker image tags to reflect user/project nomenclature

Kylep342 committed:
Implemented city and state string sanitizers

Kylep342 committed:
More docstrings for functions. Added 'LOWER' calls in SQL lookup of city/state for coordinates

Kylep342 committed:
Working on redis caching

Kylep342 committed:
Work on partitioning forecasts by date

Kylep342 committed:
Rudimentary code needed to cache/load from redis is implemented

Kylep342 committed:
Using regex to improve matching logic on 'period' arguments. Updating database model to include 'updated_at' column with auto update trigger

Kylep342 committed:
Improved message construction for forecasts

Kylep342 committed:
Merge pull request #8 from Kylep342/implement_fc_caching_via_redis

Implement fc caching via redis

Kylep342 committed:
Widened regex matches for time period inputs

Kylep342 committed:
Added LICENSE (MIT). Updating API, server, and thorcast code to serve as an API

Kylep342 committed:
Updating city formatting in lookup process. Updated README. Adding functionality to increment hit counter in postgres for a city when it has a cache hit in redis (meant to track lifetime total requests)

Kylep342 commented:
@ddrscott

Kylep342 committed:
Work on unit tests

ddrscott commented:
Might want to consider reading from stdin or command line arguments. I think it would be more friendly to use.

Look into the "click" package: https://click.palletsprojects.com/en/7.x/

ddrscott commented:
Oh, I see what you did. `run.sh` is probably not needed.

Kylep342 commented:
Yeah, neither is `build.sh` but it's less typing.

ddrscott commented:
Ha. I checked out your `dev` branch, but it seems like `master` is the one I should be reviewing.

Kylep342 commented:
Yeah, I got into a habit of committing straight to `master`, but will resume better git habits after this merge

gitter-badger committed:
Add Gitter badge

Kylep342 committed:
Merge pull request #10 from gitter-badger/gitter-badge

Add a Gitter chat badge to README.md

Kylep342 committed:
Removing build/run scripts for Docker

Kylep342 committed:
Adding retry functionality to server calls

Kylep342 committed:
Fixing typos, adding imports to thorcast.py to make error handling work

Kylep342 committed:
Added ability to recover/persist database connections

Kylep342 committed:
Adding logging

Kylep342 committed:
Fixing a typo

ddrscott commented on server.py:
```diff
@@ -0,0 +1,87 @@
+REDIS_DB = os.getenv('REDIS_DB')
+REDIS_PASSWORD = os.getenv('REDIS_PASSWORD')
+
+weather_cache = wc.WeatherCache(
```
Most Redis clients support `REDIS_URL` which would reduce the number of ENV vars you'd need to manage.

Kylep342 replied:
I'm not seeing that functionality in the Redis' official Python package. I'm not concerned about the env list. It's only a few more, and the only one likely to change is the host.

ddrscott commented on server.py:
```diff
@@ -0,0 +1,87 @@
+
+@app.route('/api/city=<city>&state=<state>', defaults={'period': 'today'})
+@app.route('/api/city=<city>&state=<state>&period=<period>')
+def lookup_forecast(city, state, period):
```
I noticed "forecast" is in a lot of method and variable names. It's a bit verbose for my taste isn't the whole app is based on forecasting.

Kylep342 replied:
Is changing it to lookup (and then having two `lookup` functions, separated by namespace) a bad idea?

ddrscott replied:
Having 2 lookups in separate modules isn't bad.

Since `server.py`, I'm not sure what else you'd lookup, so `_forecast` was redundant.
And then you have a variable named `forecast_json`, I also think that's a verbose and simply `json` would due.

Kylep342 replied:
Okay. I don't like `json` though, since that is a stdlib package. Could I just do `resp`?

ddrscott commented on test/test_utils/test_calendar.py:
```diff
@@ -0,0 +1,8 @@
+import datetime
+
+import utils.calendar as clndr
```
#nitpick I feel like `cal` is an easier abbreviation.

ddrscott commented on thorcast/forecast.py:
```diff
@@ -1,28 +1,32 @@
-
-
-def get_forecast(lat, lng):
+def forecast_from_api(lat, lng):
```
I feel like this is where naming things is important and not just nitpicking.
This should be named a single verb like `fetch` or `request`.
That way the use of the file would read like:
```python
import forecast

result = forecast.fetch(lat, lng)
```

Kylep342 replied:
Makes sense, I'll clean up names all around.

ddrscott commented on thorcast/geocode.py:
```diff
@@ -1,7 +1,21 @@
 
 
-def geocode(city, state, apikey):
+def geocode(city, state):
```
same naming critique here.
```python
import geocode

result = geocode.fetch(city, state)
# is better than
result = geocode.geocode(city, state)
```
I don't know if I'm going against the PEP grain here, but I really don't like when projects do:

```python
from geocode import geocode
result = geocode(city, state)
# or
from geocode import geocode as g
result = g(city, state)
```

Kylep342 replied:
I agree, didn't really think of it though.

ddrscott commented on thorcast/geocodex.py:
```diff
@@ -0,0 +1,112 @@
+    return f'postgresql://{username}:{password}@{host}:{port}/{db}'
+
+
+class Geocodex(object):
```
Based on the comments of what this class is doing, you're probably over doing it a bit here.
There's probable a disk cache decorator out there somewhere or you could make a simple one yourself.

The use of such a decorator would be:
```python
@disk_cache('/path/to/writable/cache/dir', ttl=86400)
def fetch_me_long_time(x, y):
   internal.request(x,y)
```

There's a good example of making your own decorator here: https://realpython.com/primer-on-python-decorators/

Kylep342 replied:
Are you saying I don't need to use Redis?

Kylep342 replied:
er, this is the Postgres wrapper class.

ddrscott commented on thorcast/thorcast.py:
```diff
@@ -1,26 +1,79 @@
+    city, state = fmts.sanitize_location(city, state)
+    key = f'{city}_{state}_{period}'.lower().replace(' ', '_')
+
+    logger.info(f'Checing Redis for forecast with key {key}')
```
`s/Checing/Checking`

ddrscott commented on thorcast/thorcast.py:
```diff
@@ -1,26 +1,79 @@
+
+    logger.info(f'Checing Redis for forecast with key {key}')
+    redis_retries = 5
+    while redis_retries:
```
probably want to make a retry contextmanager here.
Final use could look like this:
```python
with retries(5, msg='connecting to redis'):
     forecast = redis_conn.lookup(key)
```

Kylep342 replied:
How can I write a context manager that isn't connected at all to the block it manages?

ddrscott replied:
```python
import contextlib
import time

@contextlib.contextmanager
def retry(count):
    while True:
        try:
            yield
            return
        except Exception as e:
            print(f'caught exception: {e}')
            if count == 0:
                raise e
            count -=1

with retry(3):
    print('foo')
    raise 'BOOM!'
```
`RuntimeError: generator didn't stop after throw()`
Python stinks

ddrscott replied:
https://trinket.io/python3/0b51abe69e

Kylep342 replied:
Hm, it also didn't like your renegade exception.

Kylep342 replied:
https://stackoverflow.com/questions/34775099/why-does-contextmanager-throws-a-runtime-error-generator-didnt-stop-after-thro
Answer looks good there. I'll try implementing.

ddrscott replied:
I saw that before, but didn't want to get into a diatribe about it.
The `with` construct is crippled and could have been so good.
Again, Python stinks.

Kylep342 replied:
Lol ok, I'll take your word for it. Good thing you won't have to use it at work ;)

Kylep342 replied:
I'm assuming Ruby allows looping in context managers.

Kylep342 replied:
Is there a better way I can attempt a persistent connection?

Kylep342 replied:
Is there a better way I can attempt a persistent connection?

Kylep342 committed:
Renaming functions for brevity. Replacing all '%20' space escaping/unescaping with '+'.
Fixing typo in logging

Kylep342 committed:
Using ENTRYPOINT instead of CMD per Docker best practices

Kylep342 committed:
Fixing formatting bugs surrounding '+' chars as spaces

Kylep342 committed:
Updating tests

Kylep342 committed:
Made logging level an env variiable, specified at runtime
Added a debug statement for return of key from Redis due to crash I saw this morning

Kylep342 committed:
Removing discarded files from dockerignore

Kylep342 committed:
Updating Dockerfile to run tests as a part of the build

Kylep342 committed:
Removing old content generated by Cargo, since this project has no Rust

Kylep342 committed:
Ignoring pytest_cache

Kylep342 committed:
Updating README to be accurate for current project state

Kylep342 committed:
Removed commented out useless code
